lib/travis/api/v3/models/cron.rb

```diff
@@ -2,74 +2,72 @@ module Travis::API::V3
+      "weekly"  => :week,
+      "monthly" => :month
+    }
 
-    def next_enqueuing
-      if disable_by_build && last_non_cron_build_date > planned_time(LastBuild)
-        planned_time(NextBuild)
-      elsif last_cron_build_date >= planned_time(LastBuild)
-        planned_time(ThisBuild)
-      else
-        Time.now - 5.minutes
+    def schedule_next_build(from: nil)
+      # Make sure the next build will always be in the future
+      if (from && (from <= (DateTime.now.utc - 1.send(TIME_INTERVALS[interval]))))
+        from = DateTime.now.utc
       end
-    end
 
-    def planned_time(in_builds = ThisBuild)
-      case interval
-      when 'daily'
-        planned_time_daily(in_builds)
-      when 'weekly'
-        planned_time_weekly(in_builds)
-      when 'monthly'
-        planned_time_monthly(in_builds)
-      end
+      update_attribute(:next_run, (from || last_run || DateTime.now.utc) + 1.send(TIME_INTERVALS[interval]))
     end
 
-    def planned_time_daily(in_builds)
-      now = DateTime.now
-      build_today = DateTime.new(now.year, now.month, now.day, created_at.hour)
-      return build_today + 1 + in_builds if (now > build_today)
-      build_today + in_builds
+    def schedule_first_build
+      update_attribute(:next_run, DateTime.now.utc + Cron::SCHEDULER_INTERVAL)
     end
 
-    def planned_time_weekly(in_builds)
-      now = DateTime.now
-      build_today = DateTime.new(now.year, now.month, now.day, created_at.hour)
-      next_time = build_today + ((created_at.wday - now.wday) % 7)
-      return build_today + 7 * (1 + in_builds) if (now > next_time)
-      next_time + 7 * in_builds
+    def needs_new_build?
+      always_run? || (last_non_cron_build_time > (last_run || created_at))
     end
 
-    def planned_time_monthly(in_builds)
-      now = DateTime.now
-      created = DateTime.new(created_at.year, created_at.month, created_at.day, created_at.hour)
-      month_since_creation = (now.year * 12 + now.month) - (created_at.year * 12 + created_at.month)
-      this_month = created >> month_since_creation
-      return created >> (month_since_creation + 1 + in_builds) if (now > this_month)
-      created >> (month_since_creation + in_builds)
+    def skip_and_schedule_next_build
+      # Update last_run also, because this build wasn't enqueued through Queries::Crons
+      last_build_time = last_non_cron_build_time
+      update_attribute(:last_run, last_build_time)
+
+      schedule_next_build(from: last_build_time)
     end
 
-    def last_cron_build_date
-      last_cron_build = Models::Build.where(
-          :repository_id => branch.repository.id,
-          :branch => branch.name,
-          :event_type => 'cron'
-      ).order("id DESC").first
-      return last_cron_build.created_at unless last_cron_build.nil?
-      Time.at(0)
+    def enqueue
+      raise ServerError, 'repository does not have a github_id'.freeze unless branch.repository.github_id
+      unless branch.exists_on_github
+        self.destroy
```

Comment:
If branches get removed from console with code like this:
```ruby
Branch.find(1234).destroy
```
it defaults to the V2 model and doesn't pick up the dependent: :destroy so you may want to keep this as a fail safe.